api/apps/v1alpha1/nemo_datastorev2_types.go

```diff
@@ -0,0 +1,1046 @@
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+const (
+	// NemoDatastoreV2ConditionReady indicates that the NEMO datastore service is ready.
+	NemoDatastoreV2ConditionReady = "Ready"
+	// NemoDatastoreV2ConditionFailed indicates that the NEMO datastore service has failed.
+	NemoDatastoreV2ConditionFailed = "Failed"
+
+	// NemoDatastoreV2StatusPending indicates that NEMO datastore service is in pending state
+	NemoDatastoreV2StatusPending = "Pending"
+	// NemoDatastoreV2StatusNotReady indicates that NEMO datastore service is not ready
+	NemoDatastoreV2StatusNotReady = "NotReady"
+	// NemoDatastoreV2StatusReady indicates that NEMO datastore service is ready
+	NemoDatastoreV2StatusReady = "Ready"
+	// NemoDatastoreV2StatusFailed indicates that NEMO datastore service has failed
+	NemoDatastoreV2StatusFailed = "Failed"
+)
+
+// NemoDatastoreV2Spec defines the desired state of NemoDatastoreV2
+type NemoDatastoreV2Spec struct {
+	Image   Image           `json:"image,omitempty"`
+	Command []string        `json:"command,omitempty"`
+	Args    []string        `json:"args,omitempty"`
+	Env     []corev1.EnvVar `json:"env,omitempty"`
+	// The name of an secret that contains authn for the NGC NIM service API
+	AuthSecret     string                       `json:"authSecret"`
+	Labels         map[string]string            `json:"labels,omitempty"`
+	Annotations    map[string]string            `json:"annotations,omitempty"`
+	NodeSelector   map[string]string            `json:"nodeSelector,omitempty"`
+	Tolerations    []corev1.Toleration          `json:"tolerations,omitempty"`
+	PodAffinity    *corev1.PodAffinity          `json:"podAffinity,omitempty"`
+	Resources      *corev1.ResourceRequirements `json:"resources,omitempty"`
+	Expose         Expose                       `json:"expose,omitempty"`
+	LivenessProbe  Probe                        `json:"livenessProbe,omitempty"`
+	ReadinessProbe Probe                        `json:"readinessProbe,omitempty"`
+	StartupProbe   Probe                        `json:"startupProbe,omitempty"`
+	Scale          Autoscaling                  `json:"scale,omitempty"`
+	Metrics        Metrics                      `json:"metrics,omitempty"`
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:default:=1
+	Replicas     int    `json:"replicas,omitempty"`
+	UserID       *int64 `json:"userID,omitempty"`
+	GroupID      *int64 `json:"groupID,omitempty"`
+	RuntimeClass string `json:"runtimeClass,omitempty"`
+
+	DataStoreParams NemoDatastoreV2Params `json:"dataStoreParams"`
+}
+
+// NemoDatastoreV2Status defines the observed state of NemoDatastoreV2
+type NemoDatastoreV2Status struct {
+	Conditions        []metav1.Condition `json:"conditions,omitempty"`
+	AvailableReplicas int32              `json:"availableReplicas,omitempty"`
+	State             string             `json:"state,omitempty"`
+}
+
+type NemoDatastoreV2Params struct {
```

Comment:
we should split these into "db", "objectStore", "storage", "sshService" types. And map underlying values as done in the helm chart. Params itself need not be same as helm chart values.

Reply:
It's difficult to categorize the pre-existing secrets into these categorize like the NeMo datastore helmchart..
Some of the pre-existing secrets come with the helmcharts, and they don't seem to be tied to neither "db", "objectStore", "storage", nor "sshService. For example, `datastoreInitSecret` contains the shell script for initializing datastore volumes.

We can discuss further during the demo

Reply:
Another option would be to make the NeMo datastore controller fetch and deploy the nemo datastore helmchart... so that the datastore CR can be formatted exactly like the helm chart values file

Reply:
no, we will be tied to the helm spec in that case. We can further discuss on those params and see if we can avoid them while deployment.